File: libraries/apollo-runtime-java/src/main/java/com/apollographql/apollo/runtime/java/network/ws/protocol/WsProtocol.java
package com.apollographql.apollo.runtime.java.network.ws.protocol;

import com.apollographql.apollo.api.Adapters;
import com.apollographql.apollo.api.ApolloRequest;
import com.apollographql.apollo.api.CustomScalarAdapters;
import com.apollographql.apollo.api.Operation;
import com.apollographql.apollo.api.json.BufferedSinkJsonWriter;
import com.apollographql.apollo.api.json.BufferedSourceJsonReader;
import com.apollographql.apollo.runtime.java.network.ws.WebSocketConnection;
import okio.Buffer;
import okio.ByteString;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.util.Map;

public abstract class WsProtocol {
  protected WebSocketConnection webSocketConnection;
  protected Listener listener;

  public WsProtocol(WebSocketConnection webSocketConnection, Listener listener) {
    this.webSocketConnection = webSocketConnection;
    this.listener = listener;
  }

  public abstract void connectionInit();

  public abstract void handleServerMessage(Map<String, Object> messageMap);

  public abstract <D extends Operation.Data> void startOperation(ApolloRequest<D> request);

  public abstract <D extends Operation.Data> void stopOperation(ApolloRequest<D> request);


  protected void sendMessageMap(Map<String, Object> messageMap, WsFrameType frameType) {
    switch (frameType) {
      case Text:
        sendMessageMapText(messageMap);
        break;
      case Binary:
        sendMessageMapBinary(messageMap);
        break;
    }
  }

  protected void sendMessageMapText(Map<String, Object> messageMap) {
    webSocketConnection.send(toJsonString(messageMap));
  }

  protected void sendMessageMapBinary(Map<String, Object> messageMap) {
    webSocketConnection.send(toJsonByteString(messageMap));
  }

  /**
   * Receive a new WebMessage message as a `Map<String, Any?>`. Messages that aren't Json objects are ignored and the method will block
   * until the next message. Returns null if the connection is closed.
   *
   * @param timeoutMs the timeout in milliseconds or -1 for no timeout
   */
  @Nullable
  protected Map<String, Object> receiveMessageMap(long timeoutMs) {
    while (true) {
      String messageJson = webSocketConnection.receive(timeoutMs);
      if (messageJson == null) {
        return null;
      }
      Map<String, Object> map = toMessageMap(messageJson);
      if (map != null) {
        return map;
      }
    }
  }

  public void run() {
    while (true) {
      Map<String, Object> messageMap = receiveMessageMap(-1L);
      if (messageMap == null) {
        // Connection closed
        listener.networkError(new IOException("Connection closed"));
        return;
      }
      handleServerMessage(messageMap);
    }
  }

  protected static String toJsonString(Map<String, Object> messageMap) {
    Buffer buffer = new Buffer();
    BufferedSinkJsonWriter writer = new BufferedSinkJsonWriter(buffer);
    try {
      Adapters.AnyAdapter.toJson(writer, CustomScalarAdapters.Empty, messageMap);
    } catch (IOException ignored) {
    }
    return buffer.readUtf8();
  }

  protected static ByteString toJsonByteString(Map<String, Object> messageMap) {
    Buffer buffer = new Buffer();
    BufferedSinkJsonWriter writer = new BufferedSinkJsonWriter(buffer);
    try {
      Adapters.AnyAdapter.toJson(writer, CustomScalarAdapters.Empty, messageMap);
    } catch (IOException ignored) {
    }
    return buffer.readByteString();
  }

  private static Map<String, Object> toMessageMap(String messageJson) {
    try {
      //noinspection unchecked
      return (Map<String, Object>) Adapters.AnyAdapter.fromJson(new BufferedSourceJsonReader(new Buffer().writeUtf8(messageJson)), CustomScalarAdapters.Empty);
    } catch (Exception e) {
      return null;
    }
  }

  public void close() {
    webSocketConnection.close();
  }


  public interface Listener {
    /**
     * A response was received. payload might contain "errors". For subscriptions, several responses might be received.
     */
    void operationResponse(String id, Map<String, Object> payload);

    /**
     * An error was received in relation to an operation
     */
    void operationError(String id, Map<String, Object> payload);

    /**
     * An operation is complete
     */
    void operationComplete(String id);

    /**
     * A general error was received. A general error is a protocol error that doesn't have an operation id. If you have an operation id, use
     * `operationError` instead.
     */
    void generalError(Map<String, Object> payload);

    /**
     * A network error occurred A network error is terminal
     */
    void networkError(Throwable cause);
  }

  public enum WsFrameType {
    Text,
    Binary
  }

  public interface Factory {
    String getName();

    WsProtocol create(WebSocketConnection webSocketConnection, Listener listener);
  }
}

File: apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/network/ws/WsProtocol.kt
package com.apollographql.apollo3.network.ws

import com.apollographql.apollo3.annotations.ApolloInternal
import com.apollographql.apollo3.api.AnyAdapter
import com.apollographql.apollo3.api.ApolloRequest
import com.apollographql.apollo3.api.CustomScalarAdapters
import com.apollographql.apollo3.api.Operation
import com.apollographql.apollo3.api.json.BufferedSourceJsonReader
import com.apollographql.apollo3.api.json.buildJsonByteString
import com.apollographql.apollo3.api.json.buildJsonString
import com.apollographql.apollo3.api.json.writeAny
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import okio.Buffer

/**
 * A [WsProtocol] is responsible for handling the details of the WebSocket protocol.
 *
 * Implementations must implement [WsProtocol.handleServerMessage], [WsProtocol.startOperation], [WsProtocol.stopOperation]
 *
 * [WsProtocol.handleServerMessage], [WsProtocol.startOperation], [WsProtocol.stopOperation]  all share the same
 * thread and rely on [webSocketConnection] to do the operations async
 *
 * @param webSocketConnection the connection
 * @param listener a listener
 */
abstract class WsProtocol(
    protected val webSocketConnection: WebSocketConnection,
    protected val listener: Listener,
) {

  interface Listener {
    /**
     * A response was received. payload might contain "errors"
     * For subscriptions, several responses might be received.
     */
    fun operationResponse(id: String, payload: Map<String, Any?>)

    /**
     * An error was received in relation to an operation
     */
    fun operationError(id: String, payload: Map<String, Any?>?)

    /**
     * An operation is complete
     */
    fun operationComplete(id: String)

    /**
     * A general error was received
     * A general error is a protocol error that doesn't have an operation id
     * If you have an operation id, use [operationError] instead
     */
    fun generalError(payload: Map<String, Any?>?)

    /**
     * A network error occurred
     * A network error is terminal
     */
    fun networkError(cause: Throwable)
  }

  /**
   * Initializes the connection and suspends until the server acknowledges it.
   *
   * @throws Exception
   */
  abstract suspend fun connectionInit()

  /**
   * Handles a server message and notifies [listener] appropriately
   */
  abstract fun handleServerMessage(messageMap: Map<String, Any?>)

  /**
   * Starts the given operation
   */
  abstract fun <D : Operation.Data> startOperation(request: ApolloRequest<D>)

  /**
   * Stops the given operation
   */
  abstract fun <D : Operation.Data> stopOperation(request: ApolloRequest<D>)

  @OptIn(ApolloInternal::class)
  protected fun Map<String, Any?>.toByteString() = buildJsonByteString {
    writeAny(this@toByteString)
  }

  @OptIn(ApolloInternal::class)
  protected fun Map<String, Any?>.toUtf8() = buildJsonString {
    writeAny(this@toUtf8)
  }

  @Suppress("UNCHECKED_CAST")
  protected fun String.toMessageMap() = AnyAdapter.fromJson(
      BufferedSourceJsonReader(Buffer().writeUtf8(this)),
      CustomScalarAdapters.Empty
  ) as Map<String, Any?>

  protected fun sendMessageMapBinary(messageMap: Map<String, Any?>) {
    webSocketConnection.send(messageMap.toByteString())
  }

  protected fun sendMessageMapText(messageMap: Map<String, Any?>) {
    webSocketConnection.send(messageMap.toUtf8())
  }

  protected fun sendMessageMap(messageMap: Map<String, Any?>, frameType: WsFrameType) {
    when (frameType) {
      WsFrameType.Text -> sendMessageMapText(messageMap)
      WsFrameType.Binary -> sendMessageMapBinary(messageMap)
    }
  }

  protected suspend fun receiveMessageMap() = webSocketConnection.receive().toMessageMap()

  /**
   * Read the WebSocket
   */
  open suspend fun run() {
    try {
      while (true) {
        handleServerMessage(receiveMessageMap())
      }
    } catch (e: CancellationException) {
      throw e
    } catch (e: Exception) {
      listener.networkError(e)
    }
  }

  /**
   * Closes the connection gracefully.
   * It is expected that a future call to [WsProtocol.Listener.networkError] is made
   */
  open fun close() {
    webSocketConnection.close()
  }

  interface Factory {
    /**
     * The name of the protocol as in the Sec-WebSocket-Protocol header
     *
     * Example: "graphql-transport-ws" or "graphql-ws"
     */
    val name: String

    /**
     * Create a [WsProtocol]
     */
    fun create(
        webSocketConnection: WebSocketConnection,
        listener: Listener,
        scope: CoroutineScope,
    ): WsProtocol
  }
}

enum class WsFrameType {
  Text,
  Binary
}
